### ISSUE TITLE
[FEATURE] Parallelize Exact Search for vector indices in a segment

### ISSUE THREAD

**Issue by navneet1v** [opensearch-project/k-NN#2326]
## Description
Currently at a segment level when vector search is happening there are multiple cases when we move from ANN search to exact search. Cases like when filter threshold hit or if there are no graphs present on the segment due to graph creation threshold is not met for a segment. In all these cases we do exact search but while doing exact search we serially iterate over valid docIds, read the vectors and then compute the score.

## Possible Optimizations
One possible optimization we can do is to actually parallelize this, which can give a reduce the latency.

Some ideas can be:
1. We can actually gather all the docIds in an array and then let multiple threads consume those docIds with their VectorValues to read the vector and compute the scores in parallel.
2. We can also first read the vectors and docsIds in heap and then compute scores in parallel.
3. We can have a VectorValuesIterator that is thread safe and each thread can read from single iterator till the iterator reaches NO_MORE_DOCS.  Thanks to @heemin32 for suggesting this. 

I think 1 and 3, should be better over 2 as in that case we don't need to read all vectors in heap and put more stress on heap. Looking for feedbacks here. 

But during implementation we can see which one we should choose.

Similar thing can be done for the script score based exact search but that will need some more thinking. Will try to update add that too.

**Comment by heemin32** [comment]
3. We could have custom thread safe DocIdSetsIterator and let multiple thread to share the iterator to process.

**Comment by navneet1v** [comment]
> We could have custom thread safe DocIdSetsIterator and let multiple thread to share the iterator to process.

Yes we can but we need a way to tell threads from which docId to which Id it needs to do the scoring. Hence I was thinking if we just accumulate DocIds in batches and then keep on handing over to threads for scoring while we iterate over the docIds that might be easy to implement. But I like the suggestion.

**Comment by heemin32** [comment]
> > We could have custom thread safe DocIdSetsIterator and let multiple thread to share the iterator to process.
> 
> Yes we can but we need a way to tell threads from which docId to which Id it needs to do the scoring. Hence I was thinking if we just accumulate DocIds in batches and then keep on handing over to threads for scoring while we iterate over the docIds that might be easy to implement. But I like the suggestion.

The thread will process whatever available next doc id from the iterator until they meet no_more_doc.

**Comment by navneet1v** [comment]
> > > We could have custom thread safe DocIdSetsIterator and let multiple thread to share the iterator to process.
> > 
> > 
> > Yes we can but we need a way to tell threads from which docId to which Id it needs to do the scoring. Hence I was thinking if we just accumulate DocIds in batches and then keep on handing over to threads for scoring while we iterate over the docIds that might be easy to implement. But I like the suggestion.
> 
> The thread will process whatever available next doc id from the iterator until they meet no_more_doc.

Ahhh, now i know what you saying. This is pretty interesting. I think we can build such iterator with some locks and synchronized blocks. I think this will be more easy to build and validate. Thanks @heemin32 for the suggestion.

**Comment by shatejas** [comment]
@navneet1v One of the ideas I have is to use IndexInput#prefetch. Current exact search uses FlatVectors which are configured to have random access, this increases the cost of disk IO. Before attempting multithreading we should try to use prefetch and benchmark it. It needs changes in Lucene as FloatVectorValues will need a prefetch API. We can reconsider parallelizing after we benchmark prefetch. Prefetch is much more simpler to implement and maintain than multi-threading. 

I don't expect parallel searches to be affected heavily if its a controlled prefetch. Thoughts?

**Comment by navneet1v** [comment]
> [@navneet1v](https://github.com/navneet1v) [@heemin32](https://github.com/heemin32) One of the ideas I have is to use IndexInput#prefetch. Current exact search uses FlatVectors which are configured to have random access, this increases the cost of disk IO. Before attempting multithreading we should try to use prefetch and benchmark it. It needs changes in Lucene as FloatVectorValues will need a prefetch API. We can reconsider parallelizing after we benchmark prefetch. Prefetch is much more simpler to implement and maintain than multi-threading.
> 
> I don't expect parallel searches to be affected heavily if its a controlled prefetch. Thoughts?

I think prefetch is always a good optimization. But my question would be is the prefetch supported in 9.12 version of Lucene? 

Parallelizing the whole distance computation should parallelize the disk IO and distance computations. Having said that, for prefetch I would first like to know if it is supported with 9.12 version of Lucene or not.

**Comment by shatejas** [comment]
@navneet1v 9.12 won't have prefetch in float vector values, thats the change that will have to be made. I was planning to add that in lucene as a part of another fix. If this is a part of 2.19 OS release prefetch solution is not possible. But if this is a part of OS 3.0 release we probably can, assuming OS 3.x will use Lucene 10.x

I would recommend benchmarking prefetch first if this isn't targeted for 2.19 OS. Multithreading has the caveat of affecting CPU usage and can be more involved in terms of implementation

**Comment by navneet1v** [comment]
> [@navneet1v](https://github.com/navneet1v) 9.12 won't have prefetch in float vector values, thats the change that will have to be made. I was planning to add that in lucene as a part of another fix. If this is a part of 2.19 OS release prefetch solution is not possible. But if this is a part of OS 3.0 release we probably can, assuming OS 3.x will use Lucene 10.x

Good point. Right now plan is not there on when we can do this. Since this is still in exploration phase. 

> I would recommend benchmarking prefetch first if this isn't targeted for 2.19 OS. Multithreading has the caveat of affecting CPU usage and can be more involved in terms of implementation

Even to do benchmarking we need Lucene version 10. So I think, we should keep this as part of one of the enhancement and come back on this enhancement once we have Lucene version 10. But I totally agree prefetch can help here.

**Comment by buddharajusahil** [comment]
Please assign this task to me.

**Comment by navneet1v** [comment]
@shatejas lets have a separate GH issue for the prefetch thing. I think we need both of these things to improve the latencies for filters.

**Comment by navneet1v** [comment]
@buddharajusahil can we look into this PR: https://github.com/apache/lucene/pull/13542 since we have moved to lucene 10.0 and this Pr already provides the capability to have multiple DocIdSetIterator for the same segment. Can you please look into this?
